src/devices/sound/ymfm.h

```diff
@@ -0,0 +1,800 @@
+
+	// internal state
+	mutable s16 m_feedback[3];            // feedback memory for operator 1
+	ymfm_operator<RegisterType> m_op1;    // operator 1
+	ymfm_operator<RegisterType> m_op2;    // operator 2
+	ymfm_operator<RegisterType> m_op3;    // operator 3
+	ymfm_operator<RegisterType> m_op4;    // operator 4
+	RegisterType m_regs;                  // channel-specific registers
+};
+
+
+// ======================> ymfm_engine_base
+
+template<class RegisterType>
+class ymfm_engine_base
+{
+public:
+	enum : u8
+	{
+		STATUS_TIMERA = 0x01,
+		STATUS_TIMERB = 0x02,
+		STATUS_BUSY = 0x80
+	};
+
+	// constructor
+	ymfm_engine_base(device_t &device);
+
+	// register for save states
+	void save(device_t &device);
+
+	// reset the overall state
+	void reset();
+
+	// master clocking function
+	u32 clock(u8 chanmask);
+
+	// compute sum of channel outputs
+	void output(s32 &lsum, s32 &rsum, u8 rshift, s16 clipmax, u8 chanmask) const;
+
+	// write to the OPN registers
+	void write(u16 regnum, u8 data);
+
+	// return the current status
+	u8 status() const;
+
+	// set/reset bits in the status register, updating the IRQ status
+	void set_reset_status(u8 set, u8 reset) { m_status = (m_status | set) & ~reset; check_interrupts(); }
+
+	// set the IRQ mask
+	void set_irq_mask(u8 mask) { m_irq_mask = mask; check_interrupts(); }
+
+	// set the busy flag in the status register
+	void set_busy();
+
+	// return the current clock prescale
+	u8 clock_prescale() const { return m_clock_prescale; }
+
+	// set prescale factor (2/3/6)
+	void set_clock_prescale(u8 prescale) { m_clock_prescale = prescale; }
+
+	// configuration helpers
+	auto irq_handler() { return m_irq_handler.bind(); }
+
+	// reset the LFO state
+	void reset_lfo() { m_lfo_counter = 0; }
+
+private:
+	// clock the LFO, updating m_lfo_am and return the signed PM value
+	s8 clock_lfo();
+
+	// clock the noise generator
+	void clock_noise();
+
+	// update the state of the given timer
+	void update_timer(u8 which, u8 enable);
+
+	// timer callback
+	TIMER_CALLBACK_MEMBER(timer_handler);
+
+	// check interrupts
+	void check_interrupts();
+
+	// internal state
+	device_t &m_device;              // reference to the owning device
+	u32 m_env_counter;               // envelope counter; low 2 bits are sub-counter
+	u32 m_lfo_counter;               // LFO counter
+	u32 m_noise_lfsr;                // noise LFSR state
+	u8 m_noise_counter;              // noise counter
+	u8 m_lfo_am;                     // current LFO AM value
+	u8 m_status;                     // current status register
+	u8 m_clock_prescale;             // prescale factor (2/3/6)
+	u8 m_irq_mask;                   // mask of which bits signal IRQs
+	u8 m_irq_state;                  // current IRQ state
+	attotime m_busy_end;             // end of the busy time
+	emu_timer *m_timer[2];           // our two timers
+	devcb_write_line m_irq_handler;  // IRQ callback
+ 	std::vector<std::unique_ptr<ymfm_channel<RegisterType>>> m_channel; // channel pointers
```

Comment:
You can get the number of channels from the `RegisterType` as well – this doesn’t need to be a vector.

Reply:
Fair point.